### PR TITLE
Remember me functionality

### PR DESCRIPTION
**Pull request by firte07**
This pull request contains the implementation of the remember me functionality.

In this PR you can find:

- HAML file where is the new feat (views/sessions)
- SCSS website design (assets/stylesheets)
- Ruby implementation of the advanced login (app/controllers) 

You can visualize the deployed site on Heroku <a href="https://wolf-food-ordering.herokuapp.com/">here</a>.
The dashboard can be accessed by adding /products to the URL.

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -2,21 +2,26 @@ class SessionsController < ApplicationController
 
   def new; end
 
+  def login_successfully(user)
```
Comment (cristichirtos) [inline app/controllers/sessions_controller.rb:5]: You can make this method private. I suggest a clearer name for it, such as perform_log_in, or something that reflects it actually does the logging in.

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -4,17 +4,40 @@ def log_in(user)
     session[:user_id] = user.id
   end
 
-  def current_user
-    return unless session[:user_id]
+  def logged(user_id)
```
Comment (cristichirtos) [inline app/helpers/sessions_helper.rb:7]: From my point of view it's clearer if you don't separate this part of the current_user method from the rest of it. If you want to however, I recommend to have a more suggestive name for this method.

File: app/models/user.rb [hunk 3]
```diff
@@ -10,4 +12,28 @@ class User < ApplicationRecord
...
+  def User.digest(string)
+    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST : BCrypt::Engine.cost
+    BCrypt::Password.create(string, cost: cost)
+  end
+
+  def User.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
```
Comment (cristichirtos) [inline app/models/user.rb:25]: You can add a bang at the end of this method, as it modifies the internal state of the object.

File: app/models/user.rb [hunk 4]
```diff
@@ -10,4 +12,28 @@ class User < ApplicationRecord
...
+    update_attribute(:remember_digest, User.digest(remember_token))
+  end
+
+  def authenticated?(remember_token)
+    return false if remember_digest.nil?
+
+    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  end
+
+  def forget
```
Comment (cristichirtos) [inline app/models/user.rb:36]: Same as for remember.

**Review by cristichirtos — COMMENTED** [review]
Nice work! I noticed the images for the products are not loading, have you added them to Cloudinary?

File: app/controllers/sessions_controller.rb (hunk 1 above)
Comment (firte07) [reply app/controllers/sessions_controller.rb:5]: ok, thanks for it

File: app/helpers/sessions_helper.rb (hunk 2 above)
Comment (firte07) [reply app/helpers/sessions_helper.rb:7]: i changed the name of the method and also i made it private

File: app/models/user.rb (hunk 3 above)
Comment (firte07) [reply app/models/user.rb:25]: good

**Comment by firte07** [comment]
> Nice work! I noticed the images for the products are not loading, have you added them to Cloudinary?

Thanks, for now it is just the advanced authentication.

**Review by cristichirtos — APPROVED** [review]

**Review by sasaranAndrei — COMMENTED** [review]
You did a great job, but I observed that you didn't made the suggested changes in all the files.

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -5,17 +5,42 @@ def log_in(user)
...
+    forget(current_user)
     reset_session
     @current_user = nil
   end
+
+  private
+
+  def log_in_user(user_id)
+    user = User.find_by(id: user_id)
+    if user && user.authenticated?(cookies[:remember_token])
```
[the remark this answers is not in the document]
Comment (firte07) [reply app/helpers/sessions_helper.rb]: Yes, thank you.

File: app/views/sessions/new.html.haml [hunk 6]
```diff
@@ -7,6 +7,9 @@
       = f.email_field :email, class: 'form-control'
       = f.label :password
       = f.password_field :password, class: 'form-control'
+      = f.label :remember_me, class: "checkbox inline" do
```
[the remark this answers is not in the document]
Comment (firte07) [reply app/views/sessions/new.html.haml:11]: I put an empty line there. Also above the sign-up line. Thanks

File: config/cloudinary.yml [hunk 7]
```diff
@@ -0,0 +1,19 @@
+---
+development:
+  cloud_name: dlwyepfe6
+  api_key: '511413483619817'
+  api_secret: opTDXHqi_y7e-PMNSKrWf9i2KPk
```
Comment (sasaranAndrei) [inline config/cloudinary.yml:5]: The secrets should not be committed to the repo, instead you should use environment variables (via dotenv gem) or use Rails credentials

File: config/storage.yml [hunk 8]
```diff
@@ -6,6 +6,12 @@ local:
   service: Disk
   root: <%= Rails.root.join("storage") %>
 
+cloudinary:
+  service: Cloudinary
+  cloud_name: dlwyepfe6
+  api_key:    511413483619817
+  api_secret: opTDXHqi_y7e-PMNSKrWf9i2KPk
```
Comment (sasaranAndrei) [inline config/storage.yml:13]: Same for the secrets here.

File: config/cloudinary.yml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  enhance_image_tag: true
+  static_file_support: true
+test:
+  cloud_name: dlwyepfe6
+  api_key: '511413483619817'
+  api_secret: opTDXHqi_y7e-PMNSKrWf9i2KPk
+  enhance_image_tag: true
+  static_file_support: false
```
Comment (sasaranAndrei) [inline config/cloudinary.yml:19]: Nice work! Don't forget to add this secrets to .env file.

**Review by sasaranAndrei — APPROVED** [review]

File: app/helpers/sessions_helper.rb [hunk 10]
```diff
@@ -5,17 +5,42 @@ def log_in(user)
...
+    forget(current_user)
     reset_session
     @current_user = nil
   end
+
+  private
+
+  def log_in_user(user_id)
+    user = User.find_by(id: user_id)
+    if user&.user.authenticated?(cookies[:remember_token])
```
Comment (VictorMotogna) [inline app/helpers/sessions_helper.rb:41]: does the `User` has a field named `user`? Should it be `user&.authenticated?`?

File: config/cloudinary.yml [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+  api_secret: <%= ENV['CLOUDINARY_API_SECRET'] %>
+  enhance_image_tag: true
+  static_file_support: false
+production:
+  cloud_name: <%= ENV['CLOUDINARY_CLOUD_NAME'] %>
+  api_key:    <%= ENV['CLOUDINARY_API_KEY'] %>
+  api_secret: <%= ENV['CLOUDINARY_API_SECRET'] %>
+  enhance_image_tag: true
+  static_file_support: true
+test:
```
Comment (VictorMotogna) [inline config/cloudinary.yml:14]: like mentioned yesterday, the test environment is best to use the local storage and not upload images to Cloudinary

Comment (VictorMotogna) [reply config/cloudinary.yml:14]: I also noticed that the `test.rb` environment file does not change the `active_storage.service`, so I think that these credentials are not used

File: app/helpers/sessions_helper.rb (hunk 10 above)
Comment (firte07) [reply app/helpers/sessions_helper.rb]: Yes, it should. It was a mistake. I changed it

File: config/cloudinary.yml (hunk 11 above)
Comment (firte07) [reply config/cloudinary.yml]: It was a problem with the .env file from config. Looks fine now and also I have removed the 'test:` from cloudinary.yml

**Review by VictorMotogna — APPROVED** [review]

**Pull request merged**